Add "auto" provisioning type.

jp39 opened the pull request:
Automatically use hostpath if pod is on same host as zfs pool.

This addresses #85. When storage class type is set to auto, automatically create a hostpath volume when the scheduler selects the specified node to run the pod, otherwise fallback to using NFS.

Note this only works when volumeBindingMode is set to WaitForFirstConsumer in the storage class. Otherwise, when set to Immediate, volumes will be pre-provisioned before the scheduler selects a node for the pod consuming the volume, and options.SelectedNode will be unset.

Note that there could be unintended side-effects if multiple pods using the volume claim are scheduled on different nodes, depending on which pods gets scheduled first. If the first pod gets scheduled on the ZFS host, it will automatically use a hostpath volume and node affinity will be set so that other pods will be prevented from running on other nodes. On the other hand, if the second pod gets scheduled on a node which is not the ZFS host, it will use a NFS volume, and subsequent pods will also use NFS, even if scheduled to run on the ZFS host.

ccremer commented on pkg/provisioner/provision.go:
```diff
@@ -24,8 +24,10 @@ func (p *ZFSProvisioner) Provision(ctx context.Context, options controller.Provi
+	useHostPath := true
+	if parameters.Type == "nfs" || parameters.Type == "auto" && (options.SelectedNode == nil || parameters.HostPathNodeName != options.SelectedNode.Name) {
+		useHostPath = false
+		properties["sharenfs"] = parameters.NFSShareProperties
```
It's been a while since I last programmed in Go, is this not possible?
```suggestion
		properties[SharePropertiesParameter] = parameters.NFSShareProperties
```

ccremer commented on pkg/provisioner/provision.go:
```diff
@@ -77,24 +79,23 @@ func (p *ZFSProvisioner) Provision(ctx context.Context, options controller.Provi
+func createVolumeSource(parameters *ZFSStorageClassParameters, dataset *zfs.Dataset, useHostPath bool) v1.PersistentVolumeSource {
+	if !useHostPath {
 		return v1.PersistentVolumeSource{
 			NFS: &v1.NFSVolumeSource{
 				Server:   parameters.Hostname,
 				Path:     dataset.Mountpoint,
 				ReadOnly: false,
 			},
 		}
-	}
-	if parameters.HostPath != nil {
+	} else {
 		hostPathType := v1.HostPathDirectory
 		return v1.PersistentVolumeSource{
 			HostPath: &v1.HostPathVolumeSource{
```
Nit: Flip if-condition and remove else-block to reduce nesting.

ccremer commented on pkg/provisioner/provision.go:
```diff
@@ -107,9 +108,9 @@ func createVolumeSource(parameters *ZFSStorageClassParameters, dataset *zfs.Data
-	if parameters.HostPath != nil {
-		node := parameters.HostPath.NodeName
+func createNodeAffinity(parameters *ZFSStorageClassParameters, useHostPath bool) *v1.VolumeNodeAffinity {
+	if useHostPath {
```
Maybe we can also flip the if-condition here to reduce nesting:
```go
if !useHostPath {
  return nil
}
...
return &v1.VolumeNodeAffinity...
```

ccremer reviewed (commented):
Hi!
Many thanks in your interest and investing time into this.
Overall implementation looks okay, though there are some smaller improvements.

I haven't tested it yet.

Please also have a look at the test files, in CI they are failing. There is also supposed to be some integration tests, maybe you can use this to verify the intended behavior.

If you don't mind, would you write some documentation in the README? If not, I could add it later on by myself while testing, just let me know.

ccremer commented:
> Note that there could be unintended side-effects if multiple pods using the volume claim are scheduled on different nodes, depending on which pods gets scheduled first. If the first pod gets scheduled on the ZFS host, it will automatically use a hostpath volume and node affinity will be set so that other pods will be prevented from running on other nodes. On the other hand, if the second pod gets scheduled on a node which is not the ZFS host, it will use a NFS volume, and subsequent pods will also use NFS, even if scheduled to run on the ZFS host.

Hmm, I have some concerns regarding this side effect. It sounds like it'll become effectively much less deterministic regarding which type is used, especially if the first Pod lands on the ZFS host and prevents other Pods from running. This _will_ lead to issues and questions eventually.

In that sense, I don't think naming this method `auto` is good enough, as people will likely not read its "fineprint" and just use `auto` without thinking about its side effect.
If you have an idea how to control this side effect, I'm all ears. Otherwise, I'd rather prefer to name it `preferHostPathOverNfs` or something similar (better name welcome), just something that the user would then have to read the docs before choosing.

ccremer commented:
Since I've merged #126 , you should probably rebase your PR.

jp39 commented:
Thanks for your review. I've addressed most of your comments, and I rebased the branch on top of the most recent master. I still need to complete the README file. Also I'm still looking for a good enough name to replace the "auto" type.

ccremer commented on pkg/zfs/zfs.go:
```diff
@@ -114,7 +114,7 @@ func (z *zfsImpl) SetPermissions(dataset *Dataset) error {
 		return fmt.Errorf("undefined mountpoint for dataset: %s", dataset.Name)
 	}
-	cmd := exec.Command("update-permissions", dataset.Hostname, dataset.Mountpoint)
+	cmd := exec.Command("chmod", "g+w", dataset.Mountpoint)
```
Is this a leftover of testing?
Because in the docker image this wouldn't work, as updating permissions is done over SSH (see https://github.com/ccremer/kubernetes-zfs-provisioner/blob/829d4a53de247938220ff9d3310bc5205bf88106/docker/update-permissions.sh)

ccremer reviewed (commented):
Thanks for working on it!

Regarding the readme, if applicable I would refer to https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes-using-node-affinity/#schedule-a-pod-using-preferred-node-affinity to better control the side effect

jp39 commented:
> If you have an idea how to control this side effect, I'm all ears. Otherwise, I'd rather prefer to name it `preferHostPathOverNfs` or something similar (better name welcome), just something that the user would then have to read the docs before choosing.

My expertise in Kubernetes is actually still fairly limited. But after reading this https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes, it looks like pods running on multiple nodes accessing the same volume is not something that is commonly supported anyway.

I think it would be reasonable to force the NFS volume source to be used instead of HostPath, whenever the volume claim contains the `ReadOnlyMany` or `ReadWriteMany` access modes.

Then in this case, maybe it would actually make sense to call this mode "Auto".

What do you think?

ccremer commented:
You have a point regarding the access modes.
yeah, let's call it "auto", but document the behaviour in the Readme.

jp39 commented:
Hi @ccremer, I've added a few words in the README file. Let me know if that suits you. Thanks.

jp39 commented:
I also fixed lint errors and failing tests. Thanks.

ccremer commented on README.md:
```diff
@@ -15,6 +15,12 @@ but the `PersistentVolume` objects will have a [NodeAffinity][node affinity] con
 
 ![architecture with Hostpath](architecture.hostpath.drawio.svg "Architecture with Hostpath provisioning")
 
+As a third option, if the ZFS host is part of the cluster, you can let the provisioner choose
+whether [NFS][nfs] or [HostPath][hostpath] is used with the `Auto` mode. If the scheduler
+decides to place a Pod onto the ZFS host, *and* the requested access mode in the Persistent Volume
+Claim is `ReadWriteOnce` (the volume can only be accessed by pods running on the same node)
+[HostPath][hostpath] will automatically be used, otherwise [NFS][nfs] will be used.
+
```
```suggestion
As a third option, if the ZFS host is part of the cluster, you can let the provisioner choose
whether [NFS][nfs] or [HostPath][hostpath] is used with the `Auto` mode. If the scheduler
decides to place a Pod onto the ZFS host, *and* the requested access mode in the Persistent Volume
Claim contains `ReadWriteOnce` (the volume can only be accessed by pods running on the same node)
[HostPath][hostpath] will automatically be used, otherwise [NFS][nfs] will be used.

```

ccremer commented on README.md:
````diff
@@ -85,6 +91,26 @@ parameters:
 ```
 For NFS, you can also specify other options, as described in [exports(5)][man exports].
 
+The following example configures a storage class using the `Auto` type. The provisioner
+will decide whether [HostPath][hostpath] or [NFS][nfs] will be used based on where the
+pods are being scheduled.
````
Maybe I misunderstood something. Didn't we recently discuss the side-effect of scheduling, and therefore decide based on the requested Access mode instead of the scheduled node?

jp39 replied:
Do you mean we should set node affinity in the Auto mode when the volume claim contains ReadWriteOnce instead of letting the scheduler choose where to place the pods ? That's not what's currently happening in the current state of this PR but that's something we can do I suppose.

ccremer replied:
Yes, exactly.

The storage class contains the ZFS node name. If the requested Volume using this class is only RWO or RWOP, then we let the pod to be run on that node using HostPath (the volume contains the affinity rule).

However, if the volume contains RWX or ROX, then we'll let the scheduler decide, and set the volume to NFS. That way, it _could_ happen that if pods are scheduled on the same node as the ZFS node, it would still be mounted as NFS. But it also avoids the side effect, as all pods will be mounted as NFS regardless where they get scheduled

ccremer reviewed (commented):
Just a quick question, didn't do a full review yet. I hope I'll get to it soon.

jp39 commented:
Hi, I've added the changes you suggested in a second commit so you can see more easily what's been changed here.

jp39 commented on pkg/provisioner/provision.go:
```diff
@@ -73,28 +73,44 @@ func (p *ZFSProvisioner) Provision(ctx context.Context, options controller.Provi
+	case HostPath:
+		return true
+	case Auto:
+		if slices.Contains(options.PVC.Spec.AccessModes, v1.ReadOnlyMany) || slices.Contains(options.PVC.Spec.AccessModes, v1.ReadWriteMany) {
+			return false
 		}
 	}
-	if parameters.HostPath != nil {
+	return true
+}
```
There is one subtlety to take into account though: As far as I could understand it is perfectly possible to specify both ReadWriteOnce and ReadWriteMany in a volume claim. In that case, the provisionner is left to decide which of HostPath or NFS it prefers.
By inverting this condition, HostPath effectively becomes the default, whereas NFS was the default before.

jp39 replied:
Unless you really meant:

> Only return true if !ROX && !RWX

instead of:

> Only return true if RWO/RWOP, otherwise false.

Because that would make more sense to leave NFS the default.

ccremer replied:
You're right :grinning: 
It should be correct now with your latest change

ccremer commented:
Thanks for your PR! Merged it
